feat(privatebin): Add PrivateBin to incubator

schnerring opened the pull request:
**Description**
<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
-->
[PrivateBin](https://github.com/PrivateBin/PrivateBin) is a minimalist, open source online [pastebin](https://en.wikipedia.org/wiki/Pastebin) where the server has zero knowledge of pasted data.

**⚙️ Type of change**

- [x] ⚙️ Feature/App addition
- [ ] 🪛 Bugfix
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🔃 Refactor of current code

**🧪 How Has This Been Tested?**
<!--
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
-->

**📃 Notes:**
<!-- Please enter any other relevant information here -->

**✔️ Checklist:**

- [x] ⚖️ My code follows the style guidelines of this project
- [x] 👀 I have performed a self-review of my own code
- [x] #️⃣ I have commented my code, particularly in hard-to-understand areas
- [ ] 📄 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [ ] 🧪 I have added tests to this description that prove my fix is effective or that my feature works
- [ ] ⬆️ I increased versions for any altered app according to semantic versioning

schnerring committed:
Add PrivateBin to incubator

privatebin: fix linting errors

privatebin: fix privatebin.secrets definition

schnerring commented:
I defined `securityContext` in both `values.yaml` and `questions.yaml`. Will one override the other during installation of the app? Do we need both?

stavros-k commented on charts/incubator/privatebin/templates/common.yaml:
```diff
@@ -0,0 +1,5 @@
+{{/* Make sure all variables are set properly */}}
+{{- include "tc.common.loader.init" . }}
+
+{{/* Render secrets for privatebin */}}
+{{- include "privatebin.secrets" . }}
```
It's missing one more function. The apply. 
Was there any app that had it like that? (If so, we have to fix it :P)

stavros-k commented on charts/incubator/privatebin/Chart.yaml:
```diff
@@ -0,0 +1,30 @@
+dependencies:
+  - name: common
+    repository: https://library-charts.truecharts.org
+    version: 10.1.1
```
You need to add postgresql here too, otherwise it won't even deploy the dep :)

stavros-k commented on charts/incubator/privatebin/Chart.yaml:
```diff
@@ -0,0 +1,30 @@
+  - https://github.com/PrivateBin/PrivateBin
+  - https://hub.docker.com/r/privatebin/pdo  # Optimized image for database backends
+type: application
+version: 0.1.0
```
```suggestion
version: 0.0.1
```

stavros-k commented on charts/incubator/privatebin/values.yaml:
```suggestion
probes:
  liveness:
    path: "/"
  readiness:
    path: "/"
  startup:
    path: "/"
```
The default port is the port of the `main` service.

Why you want to disable startup probe?

schnerring replied:
I was just blindly copy and pasting the k8s sample from the PrivateBin docs, and they omitted the startup probe.

stavros-k replied:
Cool, generally it doesn't need to be disabled unless it's failing for some reason

stavros-k commented on charts/incubator/privatebin/values.yaml:
```suggestion
        port: 10247
```
We try to keep ports used different for every app, so users won't experience conflicts by doing the default installs
The next available port can be found here. at the bottom
https://truecharts.org/manual/default-ports
(While it says 10246, it used. the next is 10247)
We will soon have the mentioned doc automatic generated. 
(You don't have to update it)

stavros-k commented on charts/incubator/privatebin/values.yaml:
```suggestion
```
Those are the defaults so you can omit them

stavros-k replied:
You have to define podSecurityContext here, as I saw it uses different user/group based on questions.yaml

stavros-k commented on charts/incubator/privatebin/templates/_secrets.tpl:
```diff
@@ -0,0 +1,225 @@
+kind: Secret
+metadata:
+  name: privatebin-secret
+  labels:
```
I think all this file would be better fit for configmap?

stavros-k replied:
Also this file will be read-only and won't let user edit it.

If there are any options that would be frequently need to be changed, you might have to make them configurable
for example basepath

stavros-k replied:
Example on how to allow user to modify externally this file: 
https://github.com/truecharts/apps/blob/ab5ee31dc25d4cc843f09dbe801a5d0f4cce80f5/charts/stable/focalboard/templates/_configmap.tpl#L27

Check the linked app, how it does the whole linking. 
Between configmap-values-questions

schnerring replied:
> I think all this file would be better fit for configmap?

It contains sensitive Postgres connection info, so my thinking was to use a secret instead. Is there any downside to using a secret?

> If there are any options that would be frequently need to be changed, you might have to make them configurable
for example basepath

I'll add an option to configure the `basepath`. PrivateBin also allows changing PHP / nginx parameters by enqueuing additional files in the respective `conf.d` dirs. Can I do that in an additional PR before moving it to `stable`?

Also, I forgot to mount the config file 😂

stavros-k replied:
> It contains sensitive Postgres connection info, so my thinking was to use a secret instead. Is there any downside to using a secret?

I'll let @Ornias1993 for this, as I'm not entirely sure.

> I'll add an option to configure the `basepath`. PrivateBin also allows changing PHP / nginx parameters by enqueuing additional files in the respective `conf.d` dirs. Can I do that in an additional PR before moving it to `stable`?

I'd prefer to have the basic options in GUI/values.yaml so it's at least usable for people without manually adding/editing configs.
Mostly because leaving things for the future tends to not complete, and it brings unneeded bugreports/support tickets

schnerring replied:
I went ahead and exposed all the `conf.php` options to values.yaml / questions.yaml. It feels clunky having to duplicate default values in both values.yaml and questions.yaml.

I'm not sure if this was the correct approach but I required a couple of "nullable booleans". The way I did this was by declaring `string` type variables and added `true` and `false` to the `enum` field.

stavros-k replied:
> I went ahead and exposed all the `conf.php` options to values.yaml / questions.yaml. It feels clunky having to duplicate default values in both values.yaml and questions.yaml.

I agree, I have a big project in my head (and my todo list) to autogenerate the whole damn questions.yaml based on values.yaml. But sadly we can't do without `questions.yaml`. But SCALE UI needs to know what to render, and then pass changes to values.yaml.


The only way I see this changing is that they drop the UI and give access to values.yaml. But I don't think this is going to happen anytime soon

> I'm not sure if this was the correct approach but I required a couple of "nullable booleans". The way I did this was by declaring `string` type variables and added `true` and `false` to the `enum` field.

Yep, done it myself couple of times.

schnerring replied:
> I agree, I have a big project in my head (and my todo list) to autogenerate the whole damn questions.yaml based on values.yaml.

That would be amazing. I thought about how nice it would be to have a WYSIWYG side by side editor / validator.

Especially getting the "sections" right was pretty finicky (what's removable and what's required) for someone just starting out like me:

```text
# Include{persistenceBasic}
```

This isn't helm syntax right? Is this proprietary TrueNAS Web GUI syntax and documented somewhere? Is it possible to unit test the semantics of questions.yaml?

stavros-k replied:
This is just a placeholder, where a shell script replaces it before app gets pushed to `truecharts/catalog`
Here is the actual code behind it.
https://github.com/truecharts/apps/tree/master/templates/questions

Looking at those, you would realize that the questions.yaml is already a pretty slimmed down version of what it ends up looking. 
(Check a questions.yaml in catalog repo. Its huge!)

stavros-k replied:
There is a validator running for questions on the CI, but it's just code copied from iX. We don't have a clear schema.
There is some docs for it here https://github.com/truenas/charts, but from my experience there are things that missing or not clearly explained.

stavros-k commented on charts/incubator/privatebin/questions.yaml:
```diff
@@ -0,0 +1,274 @@
+                            description: "This port exposes the container port on the service"
+                            schema:
+                              type: int
+                              default: 8080
```
```suggestion
                              default: 10247
```

stavros-k commented on charts/incubator/privatebin/questions.yaml:
```diff
@@ -0,0 +1,274 @@
+            type: dict
+            attrs:
+# Include{persistenceBasic}
+# Include{persistenceAdvanced}
```
Persistence defined here must also be defined in values.yaml aswell

schnerring replied:
That storage isn't needed, I'll remove it.

stavros-k commented:
> I defined `securityContext` in both `values.yaml` and `questions.yaml`. Will one override the other during installation of the app? Do we need both?

Both files have to be in sync.

Anything not visible in the gui during installation (eg. hidden behind a checkbox). gets ignored and the value from values.yaml is used.

If its visible on the gui, that value is used instead.

stavros-k reviewed (changes requested):
In general looks nice, just little things specific to scale and/or things we haven't documented yet :)

schnerring committed:
privatebin: implement PR suggestions

Add tc.common.loader.apply

Add postgresql dependency to Chart.yaml

Change chart version to 0.0.1

Fix probes

Change service port

Omit default securityContext values

Add non-default podSecurityContext to values.yaml

Remove obsolete App Config Storage and ingress from questions.yaml

schnerring committed:
privatebin: mount secret volume

schnerring committed:
privatebin: remove fs, gcs, mysql, and sqlite from conf.php

schnerring committed:
Expose privatebin options in {values,questions}.yaml

stavros-k commented on charts/incubator/privatebin/questions.yaml:
```diff
@@ -160,38 +457,12 @@ questions:
-          schema:
-            additional_attrs: true
-            type: dict
-            attrs:
```
Why is that removed?

schnerring commented:
So how would I go about enabling users to add php.ini directives via GUI? Does a `schema.type` of `list` work? A multi-line string field type would be nice here, wouldn't it? 😎

schnerring commented:
> Error: INSTALLATION FAILED: Deployment.apps "privatebin-wdpe5vx9xe" is invalid: spec.template.spec.containers[0].volumeMounts[5].mountPath: Invalid value: "/tmp": must be unique

When adding `conf-privatebin` volume this occurs 😕 Any ideas?

stavros-k commented:
> > Error: INSTALLATION FAILED: Deployment.apps "privatebin-wdpe5vx9xe" is invalid: spec.template.spec.containers[0].volumeMounts[5].mountPath: Invalid value: "/tmp": must be unique
> 
> When adding `conf-privatebin` volume this occurs 😕 Any ideas?

I think `/tmp` is already mounted as emptyDir by default

schnerring commented:
Does this originate from the library-charts?

https://github.com/truecharts/library-charts/blob/4c2c7ddc0d87c82a55385342832aa0c3710ec073/charts/common/values.yaml#L803-L828

schnerring committed:
privatebin: remove /tmp volume

schnerring committed:
privatebin: add nginx/php sizelimit configs

schnerring committed:
privatebin: cleanup secrets conditionals

stavros-k commented on charts/incubator/privatebin/questions.yaml:
```diff
@@ -0,0 +1,544 @@
+                schema:
+                  type: boolean
+                  default: true
+                  required: true
```
required: true on booleans will allow only value of `true`.
eg. You can't set this to false.

schnerring replied:
Do I use string field + enum instead?

stavros-k replied:
If you want the user to select either false or true.
Just remove the `required` otherwise you force him to `true`

stavros-k commented on charts/incubator/privatebin/questions.yaml:
```diff
@@ -0,0 +1,544 @@
+                description: "Set a syntax highlighting theme"
+                schema:
+                  type: string
+                  default: ""
```
if you want to also have an empty option, you can add
```yaml
- value: ""
  description: "Unset"  (or empty or whatever)
```
User will see "Unset" but the actual value will be `""`

Without that option, once the user selects a value won't be able to unset it

stavros-k commented on charts/incubator/privatebin/questions.yaml:
```diff
@@ -0,0 +1,544 @@
+# Include{persistenceBasic}
+
+# Include{persistenceAdvanced}
+
```
Those should leave also, as it's part of the persistence

persistenceList should stay

schnerring committed:
privatebin: fix configMap vols

schnerring committed:
Re-add ingress; remove persistenceBasic/persistenceAdvanced

schnerring committed:
privatebin: add empty options to enums where necessary

schnerring committed:
privatebin: remove required constraints from bool questions

stavros-k commented:
> So how would I go about enabling users to add php.ini directives via GUI? Does a `schema.type` of `list` work? A multi-line string field type would be nice here, wouldn't it? 😎

Well, i'd say having fields for some basic features would be suffice. LIke max upload size, etc. The usual stuff.

If a user wants a heavy customized setup.. Might be better to fork and edit in his repo?

stavros-k commented:
> Does this originate from the library-charts?
> 
> https://github.com/truecharts/library-charts/blob/4c2c7ddc0d87c82a55385342832aa0c3710ec073/charts/common/values.yaml#L803-L828

Yep, those get injected at the start

schnerring commented:
> If a user wants a heavy customized setup.. Might be better to fork and edit in his repo?

Yes, we don't need to offer customization for this. I went ahead and calculated the appropriate values off of `.Values.privatebin.main.sizelimit`.

I think we resolved everything! This was harder than anticipated, but it was fun learning Helm, haha.

Sorry for all the noise, I hope I didn't flood you with too much email 😨

stavros-k commented:
> > If a user wants a heavy customized setup.. Might be better to fork and edit in his repo?
> 
> Yes, we don't need to offer customization for this. I went ahead and calculated the appropriate values off of `.Values.privatebin.main.sizelimit`.
> 
> I think we resolved everything! This was harder than anticipated, but it was fun learning Helm, haha.

I'll take a look again tomorrow to verify again!
You picked quite the task for your first app :P

> Sorry for all the noise, I hope I didn't flood you with too much email 😨

Nah, yours was a drop in the ocean. Getting tons of notifications for all kinds of things.
From work to fun.

Crow-Control commented:
@schnerring To be clear:
`k8s examples` do not exist!

Each and EVER helm chart community/builder has their own style. They all can (and do!) use totally different config structure for values.yaml.

stavros-k committed:
update port

stavros-k committed:
bump

stavros-k commented:
Just updated the port, as I gave you the wrong one, and bumped to today's common version

stavros-k commented:
@all-contributors please add @schnerring for code

stavros-k committed:
Merge branch 'master' into add-privatebin

Crow-Control reviewed (changes requested):
Please open a PR for the container and image.

stavros-k commented:
https://github.com/truecharts/containers/pull/3478
https://github.com/truecharts/pub/pull/51

truecharts-admin commented:
This PR is locked to prevent necro-posting on closed PRs. Please create a issue or contact staff on discord if you want to further discuss this
